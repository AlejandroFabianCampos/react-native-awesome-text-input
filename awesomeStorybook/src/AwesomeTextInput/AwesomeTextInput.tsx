import React, { FC, useRef, useState, useEffect, useCallback } from 'react'
import { Animated, TextInput, View, TextInputProps, StyleSheet, ViewStyle, TextStyle, NativeSyntheticEvent, TextInputFocusEventData } from 'react-native'

interface StyleStructure {
    container: ViewStyle,
    title: TextStyle,
    inputContainer: ViewStyle,
}

interface AwesomeTextInputProps extends TextInputProps {
    label?: string;
    customStyles?: Partial<StyleStructure>;
}

export const AwesomeTextInput: FC<AwesomeTextInputProps> = ({
    customStyles,
    label,
    ...commonTextInputProps
}) => {
    const styles: StyleStructure = {
        ...localStyles,
        ...customStyles
    }
    const [isFocused, setIsFocused] = useState(false)
    const [currentValueIsNone, setCurrentValueIsNone] = useState(true)
    const animateFocused = useRef(new Animated.Value(0)).current

    const customOnBlur = useCallback((event: NativeSyntheticEvent<TextInputFocusEventData>) => {
        setIsFocused(false);
        commonTextInputProps.onBlur && commonTextInputProps.onBlur(event);
    }, [setIsFocused, commonTextInputProps.onBlur])

    const customOnFocus = useCallback((event: NativeSyntheticEvent<TextInputFocusEventData>) => {
        setIsFocused(true);
        commonTextInputProps.onFocus && commonTextInputProps.onFocus(event)
    }, [setIsFocused, commonTextInputProps.onFocus])

    const customOnChangeText = useCallback((text: string) => {
        if (currentValueIsNone === true && text !== '') {
            setCurrentValueIsNone(false)
        } else if (currentValueIsNone === false && text === '') {
            setCurrentValueIsNone(true)
        }
        commonTextInputProps.onChangeText && commonTextInputProps.onChangeText(text)
    }, [currentValueIsNone, setCurrentValueIsNone, commonTextInputProps.onChangeText])

    useEffect(() => {
        Animated.timing(animateFocused, {
            useNativeDriver: false,
            toValue: !isFocused && currentValueIsNone ? 0 : 1,
            duration: 300
        }).start()
    }, [isFocused])

    return (
        <View style={styles.container}>
            <Animated.Text style={[
                styles.title, 
                {
                    top: animateFocused.interpolate({
                        inputRange: [0, 1],
                        outputRange: [12, -10]
                    }),
                    fontSize: animateFocused.interpolate({
                        inputRange: [0, 1],
                        outputRange: [18, 14]
                    }) 
                }
            ]}>
                {label}
            </Animated.Text>
            <View style={styles.inputContainer}>
                <TextInput
                    {...commonTextInputProps}
                    onFocus={customOnFocus}
                    onBlur={customOnBlur}
                    onChangeText={customOnChangeText}
                    testID="input"
                />
            </View>
        </View>
    )
}

const localStyles: StyleStructure = StyleSheet.create({
    container: {
        width: '100%'
    },
    inputContainer: {
        borderBottomWidth: 1,
        borderColor: '#c4c4c4',
        borderRadius: 5,
        padding: 10,
        width: '100%'
    },
    title: {
        position: "absolute",
        left: 10
    }
})
