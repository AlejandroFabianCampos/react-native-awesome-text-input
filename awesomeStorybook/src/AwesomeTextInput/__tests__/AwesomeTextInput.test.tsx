import React from "react";
import { fireEvent, render, wait } from "@testing-library/react-native";
import { AwesomeTextInput } from "../AwesomeTextInput";

jest.useFakeTimers();

test("Matches snapshot", async () => {
	const { baseElement } = render(<AwesomeTextInput />);

	expect(baseElement).toMatchSnapshot();
});

test("Custom callbacks work correctly", async () => {
	const onFocus = jest.fn();
	const onChangeText = jest.fn((text) => text);
	const onBlur = jest.fn();
	const onFocus2 = jest.fn();
	const onChangeText2 = jest.fn((text) => text);
	const onBlur2 = jest.fn();

	const { getByTestId, rerender } = render(
		<AwesomeTextInput
			onChangeText={onChangeText}
			onFocus={onFocus}
			onBlur={onBlur}
		/>
	);
	const input = getByTestId("input");
	const testText1 = "asd123";

	fireEvent.focus(input);
	expect(onFocus).toHaveBeenCalledTimes(1);
	fireEvent.changeText(input, testText1);
	expect(onChangeText).toHaveBeenCalledTimes(1);
	expect(onChangeText.mock.results[0].value).toBe(testText1);
	fireEvent.blur(input);
	expect(onBlur).toHaveBeenCalledTimes(1);

	rerender(
		<AwesomeTextInput
			onChangeText={onChangeText2}
			onFocus={onFocus2}
			onBlur={onBlur2}
		/>
	);

	fireEvent.focus(input);
	expect(onFocus2).toHaveBeenCalledTimes(1);
	fireEvent.changeText(input, testText1);
	expect(onChangeText2).toHaveBeenCalledTimes(1);
	expect(onChangeText2.mock.results[0].value).toBe(testText1);
	fireEvent.blur(input);
	expect(onBlur2).toHaveBeenCalledTimes(1);
});

test("Animates correctly without text input", () => {
	const onFocus = jest.fn();
	const onBlur = jest.fn();

	const { getByTestId } = render(
		<AwesomeTextInput onFocus={onFocus} onBlur={onBlur} />
	);
	const input = getByTestId("input");

	fireEvent.focus(input);
	expect(onFocus).toHaveBeenCalledTimes(1);
	fireEvent.blur(input);
	expect(onBlur).toHaveBeenCalledTimes(1);
});

test("Animates correctly when adding text and then removing it", () => {
	const onChangeText = jest.fn((text) => text);
	const onFocus = jest.fn();
	const onBlur = jest.fn();

	const { getByTestId } = render(
		<AwesomeTextInput
			onChangeText={onChangeText}
			onFocus={onFocus}
			onBlur={onBlur}
		/>
	);
	const input = getByTestId("input");
	const testText1 = "asd123";
	const testText2 = "";

	fireEvent.focus(input);
	expect(onFocus).toHaveBeenCalledTimes(1);
	fireEvent.changeText(input, testText1);
	fireEvent.changeText(input, testText2);
	expect(onChangeText).toHaveBeenCalledTimes(2);
	expect(onChangeText.mock.results[0].value).toBe(testText1);
	expect(onChangeText.mock.results[1].value).toBe(testText2);
	fireEvent.blur(input);
	expect(onBlur).toHaveBeenCalledTimes(1);
});
