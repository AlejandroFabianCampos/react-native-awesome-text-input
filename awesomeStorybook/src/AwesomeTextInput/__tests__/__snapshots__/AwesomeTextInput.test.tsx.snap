// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
            "left": 10,
            "position": "absolute",
            "top": 12,
          }
        }
      />
      <View
        style={
          Object {
            "borderBottomWidth": 1,
            "borderColor": "#c4c4c4",
            "borderRadius": 5,
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          rejectResponderTermination={true}
          testID="input"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`Styling matches snapshot 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "borderColor": "grey",
          "borderRadius": 10,
          "borderWidth": 1,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "white",
            "fontSize": 18,
            "left": 10,
            "position": "absolute",
            "top": 12,
          }
        }
      >
        Test label
      </Text>
      <View
        style={
          Object {
            "borderBottomWidth": 1,
            "borderColor": "#c4c4c4",
            "borderRadius": 5,
            "padding": 10,
            "width": "100%",
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          rejectResponderTermination={true}
          testID="input"
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
  </View>
</View>
`;
